### PR TITLE
 Implemented RichardHufford strat, K47R from Axelrod's Second.

### PR DESCRIPTION
**Pull request by gaffney2010**
Original here:
https://github.com/Axelrod-Python/TourExec/blob/v0.2.0/src/strategies/k47r.f

Attached some refactoring notes.  And a large set of fingerprints.

all_fortran
all_python

[RichardHuffordNotes.pdf](https://github.com/Axelrod-Python/Axelrod/files/1638140/RichardHuffordNotes.pdf)

### CODE REVIEW

File: axelrod/strategies/axelrod_second.py [hunk 1]
```diff
@@ -1597,3 +1597,116 @@ def strategy(self, opponent: Player) -> Action:
...
 
         return random_choice(self.prob_coop[number_defects])
+
+
+class RichardHufford(Player):
+    """
+    Strategy submitted to Axelrod's second tournament by Richard Hufford (K47R)
+    and came in sixteenth in that tournament.
+
+    The strategy tracks opponent "agreements," that is whenever the opponent's
```
Comment (drvinceknight) [inline axelrod/strategies/axelrod_second.py]: `"agreements",

File: axelrod/strategies/axelrod_second.py [hunk 2]
```diff
@@ -1597,3 +1597,116 @@ def strategy(self, opponent: Player) -> Action:
...
+        me_two_moves_ago = C
+        if turn > 2:
+            me_two_moves_ago = self.history[-2]
+        if me_two_moves_ago == opponent.history[-1]:
+            self.num_agreements += 1
+            self.last_four_agreements[self.last_four_index] = 1
+        else:
+            self.last_four_agreements[self.last_four_index] = 0
+
+        # Check is last_aberration is infinite.
```
Comment (drvinceknight) [inline axelrod/strategies/axelrod_second.py]: `Check if`

File: axelrod/strategies/axelrod_second.py [hunk 3]
```diff
@@ -1597,3 +1597,116 @@ def strategy(self, opponent: Player) -> Action:
...
+            self.current_streak = 0
+            return C
+
+        proportion_agree = self.num_agreements / turn
+        last_four_num = np.sum(self.last_four_agreements)
+        if proportion_agree > 0.9 and last_four_num >= 4:
+            return C
+        elif proportion_agree >= 0.625 and last_four_num >= 2:
+            return opponent.history[-1]
+        else:
```
Comment (drvinceknight) [inline axelrod/strategies/axelrod_second.py]: No need for the `else` (minor)

File: axelrod/strategies/axelrod_second.py [hunk 4]
```diff
@@ -1597,3 +1597,116 @@ def strategy(self, opponent: Player) -> Action:
...
+    def __init__(self) -> None:
+        super().__init__()
+        self.num_agreements = 2
+        self.last_four_agreements = [1] * 4
+        self.last_four_index = 0
+
+        self.streak_needed = 21
+        self.current_streak = 2
+        self.last_aberration = float("inf")
+        self.num_abb_coop = 2
```
Comment (drvinceknight) [inline axelrod/strategies/axelrod_second.py]: This is short for "number of aberrations following cooperation" right? In which case I think the `abb` is confusing, perhaps:

```
ab_after_coop_count
ab_after_def_count
```

?

Happy for other suggestions.

File: axelrod/strategies/axelrod_second.py [hunk 5]
```diff
@@ -1550,7 +1550,7 @@ def strategy(self, opponent: Player) -> Action:
 class Black(Player):
     """
     Strategy submitted to Axelrod's second tournament by Paul E Black (K83R)
-    and came in fourteenth in that tournament.
+    and came in fifteenth in that tournament.
```
Comment (drvinceknight) [inline axelrod/strategies/axelrod_second.py]: As an FYI, this is in essence stepping on #1161. No problem at all (don't change it) but just a headsup.

File: axelrod/strategies/axelrod_second.py (hunk 4 above)
Comment (gaffney2010) [reply axelrod/strategies/axelrod_second.py]: It's cooperations after aberration, so maybe coop_after_abb_count

Comment (drvinceknight) [reply axelrod/strategies/axelrod_second.py]: Cool, sounds good but a single `b` (not `abb`?): `coop_after_ab_count` and `def_after_ab_count`?

Comment (gaffney2010) [reply axelrod/strategies/axelrod_second.py]: Because I can't spell.  A single `b` is good.

File: axelrod/strategies/axelrod_second.py (hunk 1 above)
Comment (gaffney2010) [reply axelrod/strategies/axelrod_second.py]: I'm not really sure what you're asking for with this one?

Comment (drvinceknight) [reply axelrod/strategies/axelrod_second.py]: Tiny typo I believe: you currently have the comma (`,`) inside the quotation marks (`"agreements," that`), I believe it should be after the closing quotation mark (`"agreements", that`).

**Comment by drvinceknight** [comment]
Thanks! :+1:

**Pull request merged**